### DISCUSSION TITLE
2.4.0 - SerialUPDI compatibility fixes, oscillator tuning, and much more!

### DISCUSSION THREAD

**Discussion by SpenceKonde** [SpenceKonde/megaTinyCore#503] [Future Development Discussion]
megaTinyCore 2.4.0 is FINALLY HERE!!!
This brings what is hoped will be a major improvement in compatibility of SerialUPDI (except in turbo mode - turbo mode is slowed significantly by those improvements; turbo mode is not expected to work on all adapters or under all conditions, but rather to offer the absolute highest performance, with the expectation that the user will take care to provide optimum conditions)
We also see some exciting new features:
 * Tuned Internal Oscillator - see the tuning document for more information, this gives a route to overclocking these parts without using an external clock, as well as, on Optiboot configurations, running at 16 MHz when the bootloader was burned at 20 MHz and vise versa. Running the tuning sketch first is strongly recommended. The 2-series parts, at 5v and room temperature *will often run at 32 MHz* from the internal oscillator (after cranking the calibration way the hell up)! Under similar conditions, 0/1-series parts can sometimes run stabily at 30 MHz from internal, and 24 or 25 MHz should be no problem at all. Obviously, the operating temperature range will be more limited when overclocked in this way (and the internal oscillator can never push the parts quite as far as an external clock can). Be sure to read the tuning documentation if you're interested in using this functionality. 
 * New advanced digital I/O functions: openDrain(), openDrainFast() and pinConfigure (first introduced on DxCore) are now available. The first two just provide a wrapper around writing to the direction bit instead of the output bit, allowing it to be used as an "open drain" output, which is very useful for interacting with devices running at a lower voltage, while pinConfigure() provides a single function to set (or leave unchanged) every parameter of the specified pin.
 * Early version of Event.h - however, my initial work with it showed that there are still serious features needed to make this particularly useful. We're getting there though. 
 * C++ 17 support and support for libraries that insist on providing only precompiled source code. 

 There's also a HUGE number of minor and major bugfixes in here this release. Truly too many to list - the changelog is not complete, and even that runs to over 30 bullet-points. 

 We also now have better CI testing to catch any future problems which result in failures to compile, and look forward to improving this testing further.

<hr /><em>This discussion was created from the release <a href='https://github.com/SpenceKonde/megaTinyCore/releases/tag/2.4.0'>2.4.0 - SerialUPDI compatibility fixes, oscillator tuning, and much more!</a>.</em>

**Comment by hmeijdam** [comment]
I think that some congratulations are appropriate. Proficiat and thanks, as I suspect it was a boatload of work to get this out of the door.

**Comment by SpenceKonde** [comment]
Thanks. Take this boatload out for a shakedown cruise, find any huge bugs, because I need to do a 2.4.1 this weekend, because I discovered about 20 minutes after releasing.... that SerialUPDI was missing device definitions for the 4k and 8k 2-series parts! So that has to get fixed along with a huge bug in openDrainFast() (address offset by 20 instead of 0x20 per port. so if used on ports other that port A, it scribbles over some other port-related register. 

I am just stunned by how well these bad boys overclock! I've yet to find a 2-series that can't do 32 MHz from internal cranked up to near the maximum value @5V/25C!

**Comment by hmeijdam** [comment]
I see 2.4.1. is the latest release now and I am wondering if it will be also available as Arduino IDE install-able release. Or only as a manual install?

**Comment by SpenceKonde** [comment]
2.4.1 was bad. 2.4.2 will be released within a few hours.

____________
Spence Konde
Azzy’S Electronics

New products! Check them out at tindie.com/stores/DrAzzy
GitHub: github.com/SpenceKonde
ATTinyCore: Arduino support for almost every ATTiny microcontroller
Contact: ***@***.***
